includes/downloads/recalculations.php

```diff
@@ -55,31 +64,41 @@ function edd_maybe_schedule_download_recalculation( $download_id ) {
  * @return void
  */
 function edd_recalculate_order_item_download( $order_item_id, $data = array(), $previous_order_item = false ) {
+
 	// Recalculations do not need to run when the order item is first being added to the database if it's pending.
 	if ( 'edd_order_item_added' === current_action() && ( empty( $data['status'] ) || 'pending' === $data['status'] ) ) {
 		return;
 	}
+
 	// If the order item data being updated doesn't affect sales/earnings, recalculations do not need to be run.
 	if ( $previous_order_item instanceof EDD\Orders\Order_Item ) {
-		$columns_affecting_stats = array( 'status', 'quantity', 'total', 'subtotal' );
+		$columns_affecting_stats = array( 'status', 'quantity', 'total', 'subtotal', 'discount', 'tax', 'rate', 'product_id' );
+
 		// If the data being updated isn't one of these columns then we don't need to recalculate.
 		if ( empty( array_intersect( array_keys( $data ), $columns_affecting_stats ) ) ) {
 			return;
 		}
+
 		// If the data exists but matches, we don't need to recalculate.
 		if (
-			! empty( $data['status'] ) && $previous_order_item->status === $data['status'] &&
-			! empty( $data['quantity'] ) && $previous_order_item->quantity === $data['quantity'] &&
-			isset( $data['total'] ) && $previous_order_item->total == $data['total'] &&
-			isset( $data['subtotal'] ) && $previous_order_item->subtotal == $data['subtotal']
+			( empty( $data['status'] ) || $previous_order_item->status === $data['status'] ) &&
+			( empty( $data['quantity'] ) || $previous_order_item->quantity === $data['quantity'] ) &&
+			( ! isset( $data['total'] ) || $previous_order_item->total == $data['total'] ) &&
+			( ! isset( $data['subtotal'] ) || $previous_order_item->subtotal == $data['subtotal'] ) &&
+			( ! isset( $data['discount'] ) || $previous_order_item->discount == $data['discount'] ) &&
+			( ! isset( $data['tax'] ) || $previous_order_item->tax == $data['tax'] ) &&
+			( ! isset( $data['rate'] ) || $previous_order_item->rate == $data['rate'] ) &&
+			( empty( $data['product_id'] ) || $previous_order_item->product_id == $data['product_id'] )
 			) {
 			return;
 		}
 	}
+
 	$order_item = edd_get_order_item( $order_item_id );
 	if ( empty( $order_item->product_id ) ) {
 		return;
 	}
+
 	edd_maybe_schedule_download_recalculation( $order_item->product_id );
```

Comment:
If the `$order_item->product_id` changes, we probably need to trigger an update on that old `product_id` as well.